digital_land/package/sqlite.py

```diff
@@ -46,18 +49,100 @@ def spatialite(self, path=None):
+            self.connection = sqlite3.connect(self.path)
 
         if self._spatialite:
             self.connection.enable_load_extension(True)
             self.connection.load_extension(self._spatialite)
 
     def disconnect(self):
         logging.debug("sqlite3 disconnect")
+        if self.cursor:
+            self.cursor.close()
+            self.cursor = None
+
         self.connection.close()
 
+    def create_cursor(self, optimised=False):
+        if self.cursor:
+            self.cursor.close()
+            self.cursor = None
+
+        self.cursor = self.connection.cursor()
+        if optimised:
+            self.cursor.execute("PRAGMA synchronous = OFF")
+            self.cursor.execute("PRAGMA journal_mode = OFF")
+            self.cursor.execute("BEGIN")
+
+    def commit(self):
+        logger.debug("committing ..")
+
+        self.connection.commit()
+
+    def execute(self, cmd):
+        logger.debug(cmd)
+        try:
+            self.cursor.execute(cmd)
+        except sqlite3.Error as error:
+            logging.error("Exception: %s" % (error.__class__))
+            logging.error("Sqlite3 error: %s" % (" ".join(error.args)))
+            logging.info("Sqlite3 cmd: %s" % (cmd))
+            sys.exit(3)
+
+    def colvalue(self, row, field):
+        value = str(row.get(field, ""))
+        t = self.field_coltype(field)
+        if t == "INTEGER":
+            if value == "":
+                return "NULL"
+            return "%d" % Decimal(value)
+        if t == "JSON":
+            if value == "{}":
+                return "NULL"
+
+        return "'%s'" % value.replace("'", "''")
+
+    def colvalue2(self, row, field):
+        value = str(row.get(field, ""))
+        return value.replace("'", "''")
+
+    def insert(self, table, fields, upsert=False):
+        fields = [field for field in fields if not field.endswith("-geom")]
+
+        table_cols = ",".join([colname(field) for field in fields])
+        placeholders = ",".join(["?"] * len(fields))  # self.bindings_list(fields)
+        table = colname(table)
+
+        if upsert:
+            condition = " ON CONFLICT DO NOTHING"
+            sql_stmt = f"INSERT OR REPLACE INTO {table}({table_cols}) VALUES({placeholders}){condition};"
+        else:
+            sql_stmt = (
+                f"INSERT OR REPLACE INTO {table}({table_cols}) VALUES({placeholders});"
+            )
+
+        try:
+            self.cursor.executemany(sql_stmt, self.table_values)
+
+        except sqlite3.Error as error:
+            logging.error("Exception: %s" % (error.__class__))
+            logging.error("Sqlite3 error: %s" % (" ".join(error.args)))
+            logging.info("Sqlite3 cmd: %s" % (sql_stmt))
+            sys.exit(3)
+
+    def add_table_values(self, fields, row):
+        fields = [field for field in fields if not field.endswith("-geom")]
+
+        table_vals = tuple([self.colvalue2(row, field) for field in fields])
+        self.table_values.append(table_vals)
```

Comment:
I had the same concern.
I ran some memory tests using listed-building, which has a transformed size of 652MB.
The memory size of table values array was only in the 30s or 40s of MB.
That said, I'm definitely open to testing in a non-prod environment to check such things.
It should be possible to periodically commit a batch of say 10000 records, for example.

Reply:
Listed buildings has points for geometries though where as flood risk zones has full geometries I can imagine this impacts it. Also with the inspire polygons on the horizon introducing more memory based restrictions seems dangerous. At the minimum we need to implement batching even if it's a large batch for now. I'm wondering if it needs to be saved in memory here or is it can be directly loaded into the transaction.

Reply:
I'm leaning toward batching, otherwise we lose the benefit of batch transactions.
We'd just be back to where we started, effectively.
During my analysis, this part has the greatest potential to increase performance.

That said, if we're getting into the realms of processing GB file sizes, we might need to at least start thinking about more robust, enterprise-grade solutions.
You know what they say about rolling your own :).

Anyway, if you're not overly opposed to the current solution being batched, then I'll implement that.

Thanks for the heads up re. FRZ.
I was under the impression listed-building was the principal problem ds, and everything else would only be smaller.